Code Review

scottlittle18 opened the pull request.

scottlittle18 committed:
Re-adding scripts folder for code review

dantogno commented on Assets/Scripts/Checkpoint.cs:
```diff
@@ -0,0 +1,47 @@
+using UnityEngine;
+using UnityEngine.Animations;
+
+public class Checkpoint : MonoBehaviour {
```
Curly brackets on new line 
```
public class Checkpoint
{
```

dantogno commented on Assets/Scripts/Checkpoint.cs:
```diff
@@ -0,0 +1,47 @@
+    [SerializeField]
+    public bool isActivated = false;
+   
+    public Animator anim;
```
avoid public fields. Either use SerializeField if all you need to do is expose to editor, or use public Properties to expose variables to other classes. 
* https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/properties
* https://unity3d.com/learn/tutorials/topics/scripting/properties

dantogno commented on Assets/Scripts/Checkpoint.cs:
```diff
@@ -0,0 +1,47 @@
+   
+    public Animator anim;
+
+    //creates a variable that refers to the PlayerController
```
Remove comment, it's self explanatory here.

dantogno commented on Assets/Scripts/Checkpoint.cs:
```diff
@@ -0,0 +1,47 @@
+
+    private void Start()
+    {
+        anim = this.GetComponent<Animator>();       
```
the `this` here is not necessary. Normally studios will either always use it (for clarity) or only use it when it's necessary (for brevity). I guess I normally only use it when necessary. Pick a standard and stick to it. I guess since we're in my class I'd say don't use it unless it's necessary.

scottlittle18 replied:
I'm usually the same, my programming teacher always uses it and I've been trying to keep it from rubbing off on me but I guess it slipped that time

scottlittle18 replied:
I'm usually the same, my programming teacher always uses it and I've been trying to keep it from rubbing off on me but I guess it slipped that time

dantogno commented on Assets/Scripts/Checkpoint.cs:
```diff
@@ -0,0 +1,47 @@
+
+public class Checkpoint : MonoBehaviour {
+    [SerializeField]
+    public bool isActivated = false;
```
you don't want to make public fields also serialized fields, because that's redundant. But also, I don't want public fields period. You actually have a good example here. The preferred way is to use a Property (which is just a fancy, built-into-the-language feature for creating methods like you have `SetAsActivated` below.

It's common to want to do stuff every time a value changes. This is why in C++ you make methods like you have here, and in C# you use Properties, because that's what they were made for (since this is so common).

Imagine another developer is working on your game and doesn't realize they are supposed to use `SetAsActivated`, and just go ahead and directly set `isActivated`. It's a likely scenario, since `isActivated` is exposed as part of the public API for `CheckPoint`.

Instead, we want to make `isActivated` private so this never happens. Then if you need it exposed to the editor, you mark it as a serialized field. Then you use a property (or a method like you have here) as the public API for changing isActivated, and since it's the only choice that sounds right that you can see from other classes, that's what people will use.

scottlittle18 replied:
Oops, I think that was before I started using TODO comments to mark what I've done to debug so I can remove it later

dantogno commented on Assets/Scripts/Checkpoint.cs:
```diff
@@ -0,0 +1,47 @@
+        }
+    }
+
+    public void SetAsActivated(bool value)
```
you could try making this a property and put these lines in the "setter" of the property.

dantogno commented on Assets/Scripts/CreditScroller.cs:
```diff
@@ -0,0 +1,17 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class CreditScroller : MonoBehaviour {
```
Class does nothing, so remove it!

scottlittle18 replied:
I created it since I'm going to be using it for the credits

dantogno commented on Assets/Scripts/Door.cs:
```diff
@@ -0,0 +1,36 @@
+
+public class Door : MonoBehaviour
+{
+    BoxCollider2D box;
```
explicitly specify `private`

dantogno commented on Assets/Scripts/Door.cs:
```diff
@@ -0,0 +1,36 @@
+{
+    BoxCollider2D box;
+    
+    GameObject[] EnemyBandits;
```
explicitly specify `private` (do this for all of the private fields and methods).

dantogno commented on Assets/Scripts/Door.cs:
```diff
@@ -0,0 +1,36 @@
+
+    void OnTriggerStay2D(Collider2D other)
+    {
+        if ((Input.GetKey(KeyCode.UpArrow) || Input.GetKey(KeyCode.W)) && other.tag == "Player")
```
Use the input manager (`Input.GetButton`  or `Input.GetAxis` and related methods)

dantogno replied:
i.e., don't use input.getkey

dantogno commented on Assets/Scripts/Door.cs:
```diff
@@ -0,0 +1,36 @@
+        {
+            foreach (GameObject Enemy in EnemyBandits)
+            {
+                Enemy.GetComponent<Rigidbody2D>().constraints = RigidbodyConstraints2D.FreezePositionX;
```
Why do we do this? Extract the `foreach` loop in it's own named method and add a summary tag that explains what it's for. https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/summary

dantogno replied:
hotkey in visual studio for summaries: type /// into the text editor

dantogno commented on Assets/Scripts/EnemyHealth.cs:
```diff
@@ -0,0 +1,57 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class EnemyHealth : MonoBehaviour {
```
Move all {'s onto next line

dantogno replied:
I'd add a summary tag before the class declaration itself to explain what this is for. https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/summary

dantogno commented on Assets/Scripts/EnemyHealth.cs:
```diff
@@ -0,0 +1,57 @@
+
+public class EnemyHealth : MonoBehaviour {
+
+    //Declare variables for the different states of health
```
no need for comment here. explicitly write `private` though!

dantogno commented on Assets/Scripts/EnemyHealth.cs:
```diff
@@ -0,0 +1,57 @@
+        fullHealth = transform.GetChild(0);
+        halfHealth = transform.GetChild(1);
+        noHealth = transform.GetChild(2);
+
```
remove extra white space

dantogno commented on Assets/Scripts/EnemyHealth.cs:
```diff
@@ -0,0 +1,57 @@
+	void Start () {
+        fullHealth = transform.GetChild(0);
+        halfHealth = transform.GetChild(1);
+        noHealth = transform.GetChild(2);
```
magic numbers 😱  what are they?

dantogno commented on Assets/Scripts/EnemyHealth.cs:
```diff
@@ -0,0 +1,57 @@
+	
+	// Update is called once per frame
+	void Update () {
+        
```
remove unused methods

dantogno commented on Assets/Scripts/EnemyHealth.cs:
```diff
@@ -0,0 +1,57 @@
+
+        EnemyController = GetComponentInParent<EnemyPatrol>();
+        health = EnemyController.enemyHealth;
+
```
remove extra white space

dantogno commented on Assets/Scripts/EnemyHealth.cs:
```diff
@@ -0,0 +1,57 @@
+        
+	}
+
+    public void UpdateHealth(short currentHealth)
```
another situation where a property might work out for you.

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+
+public class EnemyPatrol : MonoBehaviour
+{
+
```
remove extra white space

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+{
+
+    [SerializeField]
+    private float moveSpeed, checkRadius, attackDelay, deathDelay, knockbackPower, knockbackHeight, knockbackTime;
```
These are great names but I'm gonna be annoying and ask you to break this up into individually declared serialized fields with ToolTip attributes to help clarify. https://docs.unity3d.com/ScriptReference/TooltipAttribute.html

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+    [SerializeField]
+    private float moveSpeed, checkRadius, attackDelay, deathDelay, knockbackPower, knockbackHeight, knockbackTime;
+
+    public short enemyHealth;
```
I think it's weird to have an `enemyHealth` variable that's not of the type `EnemyHealth`. Also, no public fields!

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+
+    CircleCollider2D enemyAttackPoint;
+
+    [SerializeField]
```
Group all the serialized fields at the top, wrap them in a region labeled "serialized fields" https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives/preprocessor-region  

https://docs.google.com/document/d/1yrKCY_sVAdWkwgWAjJLiJFj1OOHn6dC__pGkSVB61YY/edit#heading=h.oezlj2pye3ju

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+        SetupObject();
+    }
+
+    //Gather and Set Necessary Components, GameObjects, Values, and States
```
Change this comment to a summary tag. A lot of times you might call this type of function `Initialize()`

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+            enemyAttackPoint.enabled = false;
+    }
+
+    void UpdatePM()
```
what's a PM? Remove abbreviation, add summary tag.

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+    //Handle movement of Enemies to let them 'patrol' the area
+    private void Patrol()
+    {
+        if (hittingWall || notEdge)
```
What's not edge?

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+    private void Patrol()
+    {
+        if (hittingWall || notEdge)
+            moveRight = !moveRight;
```
remember, we want these to be yes or no questions. E.g. `isMovingRight`

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+            transform.localScale = new Vector3(-1, 1, 1);
+            enemyRigidBody.velocity = new Vector2(-moveSpeed, enemyRigidBody.velocity.y);
+        }
+
```
remove extra white space

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+    private void PositionChecks()
+    {
+        //WALL CHECK
+        hittingWall = Physics2D.OverlapCircle(pathCheck.position,
```
I'd extract each of these into their own methods like `UpdateIsHittingWall()` with summaries that explain what they do.

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+            }
+
+            isDamagable = false;
+
```
remove white space

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+            EnemyKnockback();
+            StartCoroutine(KnockbackTime());
+            //Check Enemy Health
+            if (enemyHealth == 0)
```
Maybe put this in method called `CheckEnemyHealth`

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+        }
+    }
+
+    // Determine Knockback Direction
```
replace comment with summary

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+    // Determine Knockback Direction
+    void EnemyKnockbackDirection(Collider2D collision)
+    {
+        //TODO: Add knock back to enemy
```
still need to do?

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+        canMove = true;
+    }
+    //------------------------------------------------------------------<<<<--------ENEMY MELEE CHECK-----------
+    void MeleeCheck()
```
add summary

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+        hasAttacked = false;
+        isAttacking = false;
+    }
+
```
only one line between methods

dantogno commented on Assets/Scripts/EnemyPatrol.cs:
```diff
@@ -0,0 +1,239 @@
+    }
+
+
+    //Remove Object When Player Kills It
```
convert comment to summary

dantogno commented on Assets/Scripts/MenuButtons.cs:
```diff
@@ -0,0 +1,41 @@
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+public class MenuButtons : MonoBehaviour {
```
new line { (you can change the default unity script template, but it's kind of annoying if you work with lots of different versions and computers)

dantogno commented on Assets/Scripts/MenuButtons.cs:
```diff
@@ -0,0 +1,41 @@
+        currentScene = SceneManager.GetActiveScene().buildIndex;
+	}
+	
+    public void PressStartButton()
```
This should probably be past tense if it is meant to be called when the player pressed the start button. It sounds like it presses the start button for you in present tense.

dantogno commented on Assets/Scripts/MenuButtons.cs:
```diff
@@ -0,0 +1,41 @@
+
+    public void Quit()
+    {
+        //TODO: Debug.Log("Quit Program Input Accepted")
```
still todo?

scottlittle18 replied:
I'm using it to make sure that I don't forget to remove the debug statement that let's me test the quit/exit button in the editor, once I finalized the art and finished the implementation of the main menu I was going to remove it

dantogno commented on Assets/Scripts/SceneLoader.cs:
```diff
@@ -0,0 +1,23 @@
+
+    public void Retry()
+    {
+
```
remove unused methods

dantogno reviewed (changes requested):
Wow great work! I have some minor notes I'd like you to address, but I'm impressed by what you've done! 🥇 Please prioritize so I can have you review one of your classmates ASAP :)

scottlittle18 committed:
Remove Empty Space, Switch some variables into Properties, Refactered To Add or Remove Methods from Scripts

- Removed most (if not all) of the empty space in my scripts
- Added Regions to PlayerController and EnemyPatrol scripts to group serialized and unserialized vairable declarations
- Changed some variables into properties
- Moved curly braces to new lines
- Clearly labeled the accesibility of all declared variables
- Added HalthEnemyMovement() method to the door script

**Still running into problems implementing the Property for the checkpoint script so the animations update properly

scottlittle18 committed:
Move curly braces to new line

- Missed a few curly braces in the last commit

scottlittle18 committed:
Simplify If statement conditions

scottlittle18 committed:
Fix Checkpoint System

- Fixed the issues with the checkpoint property of the Checkpoint script
- Refactored the enumerators and properties into their own regions

scottlittle18 committed:
Clarify Variable Names, Reorganize code

- Renamed the physics material variables to more clearly identify them as such
- Grouped the enumerators in the EnemyPatrol script into their own region

dantogno commented on Assets/Scripts/Checkpoint.cs:
```diff
@@ -3,22 +3,34 @@
     private void Update()
-    {
+    {        
         UpdateAnimation();
```
Do you need to call this every frame? If it only needs to be called when IsActivated is set, you can call it from IsActivated's setter.

scottlittle18 replied:
I tried and for some reason it didn't want to work, it kept getting null reference exceptions and other bugs

dantogno commented on Assets/Scripts/EnemyHealth.cs:
```diff
@@ -1,53 +1,60 @@
 
-        EnemyController = GetComponentInParent<EnemyPatrol>();
-        health = EnemyController.enemyHealth;
+        //EnemyController = GetComponentInParent<EnemyPatrol>();
```
Why is this commented out? Ideally there is no commented out code.

scottlittle18 replied:
That was my bad, I caught and fixed it about a minute after the last push

dantogno commented on Assets/Scripts/EnemyHealth.cs:
```diff
@@ -1,53 +1,60 @@
+        set
+        {
+            currentEnemyHealth = value;
+            UpdateHealthSprite();
```
Excellent use of property!

scottlittle18 committed:
Remove Unnecassary Comments

- Removed commented out code from the EnemyHealth script

scottlittle18 committed:
Improve Checkpoint System

- Improved the checkpoint system by updating the animation of the checkpoints in the setter or the property
- Removed Update() from the Checkpoint script and add Awake() instead

scottlittle18 committed:
Begin building level 1

- Began building Level 1

scottlittle18 committed:
Add Main Menu Button Sprites

- import main menu sprites
